### ISSUE TITLE
Comparison between strings and ints is allowed and should not be

### ISSUE THREAD

**Issue by LiterallyUniqueLogin** [pola-rs/polars#3170]
#### What language are you using?

**Python**

#### What version of polars are you using?

0.13.23

#### What operating system are you using polars on?

Rocky Linux 8.5 (Green Obsidian)

#### What language version are you using
3.7.12

#### Describe your bug.

When you try to compare a string and an integer, polars produces incorrect results.

#### What are the steps to reproduce the behavior?

```
pl.DataFrame({'a': [1.1,21,31,21,51,61,71,81]}).select(pl.col('a') < '9')
```

#### What is the actual behavior?

```
shape: (8, 1)
┌──────┐
│ a    │
│ ---  │
│ bool │
╞══════╡
│ true │
├╌╌╌╌╌╌┤
│ true │
├╌╌╌╌╌╌┤
│ true │
├╌╌╌╌╌╌┤
│ true │
├╌╌╌╌╌╌┤
│ true │
├╌╌╌╌╌╌┤
│ true │
├╌╌╌╌╌╌┤
│ true │
├╌╌╌╌╌╌┤
│ true │
└──────┘
```

#### What is the expected behavior?

It should act similarly to how native python acts when you compare a string and an integer - fail:

```
>>> '1.1' < 9
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'str' and 'int'
```

**Comment by ritchie46** [comment]
We coerce both supertypes that allow completing the operation. In this case both are converted to string values. 

Python also allows comparing string to string operations:

```python
assert not "a" < "9"
```

**Comment by SimonSchneider** [comment]
One thing to consider here is that to a lot of users this is probably a "feature" and not "unintended behaviour". I'd guess that a non negligible set of users are relying on this feature.
I'm not personally affected, we do our own validation of type coercion before building a polars query so I have no real stake in this question.
But if this is to be changed I would vote for allowing both behaviours and hiding one behind a flag or having different functions for coercing and non coercing filters.

**Comment by cnpryer** [comment]
For this example imo I’d rather it error out.

**Comment by LiterallyUniqueLogin** [comment]
@ritchie I agree Python allows you to compare strings to each other, it
just is meaningless in the context of what I intended.

I feel following language precedent makes sense. If JavaScript allows such
conversions automatically, then do so there. But not in python where people
don't expect it (and in fact some choose to work in python so they don't
have to deal with those headaches)

**Comment by ritchie46** [comment]
> @Ritchie I agree Python allows you to compare strings to each other, it just is meaningless in the context of what I intended. I feel following language precedent makes sense. If JavaScript allows such conversions automatically, then do so there. But not in python where people don't expect it (and in fact some choose to work in python so they don't have to deal with those headaches)

Note that polars is intended as a DataFrame backend for many languages. Python is definitely not guiding here.

I understand this not what you intended to do and I  maybe considering changing the behavior. But I think we should come up with rationale for polars as a own DSL, not the language you call it from.

**Comment by cnpryer** [comment]
Is the only motivation to do this to prevent errors from occurring where conversion is not handled and should be? I'm trying to understand the rationale for doing it.

**Comment by slonik-az** [comment]
> We coerce both to supertypes that allow completing the operation. In this case both are converted to string values.

**TL;DR:** Converting a number to a string and then comparing strings leads to nasty data dependent bugs and unexpected behavior (see below for details). It would be better to convert string to a number and then compare numbers. It will be faster as well (one `str->num` conversion and comparing numbers instead of `series.length()`  `num->str` conversions followed by string comparisons that are slower).

## Details 
I will be using Python to show some cases where converting a number to a string and comparing with another string leads to very strange results:

`str(1E11) > '2'   # ==> False`   -- really?

```python
from math import nan
str(nan) == nan  # ==> True  # nan is not equal to any number including itself
```

`str(123) < "I should not be comared to numbers" # ==> True` - no sanity checks of the string in the condition.

BTW, Python does not allow mixing numbers and strings in `<` or `>`  comparisons:
`1 < "9" # ==> TypeError: '<' not supported between instances of 'int' and 'str'`

If a certain target language for Polars bindings (Python or Javascript) allows for comparisons that make little sense from the Data Science point of view, Polars should stick with what is good for Data Science rather than for a target language especially since there are several of them with different rules. 

### Currently in Polars
```python
import polars as pl
pl.version()  # ==> '0.13.23'
pl.DataFrame({'a': [1.1,21,31]}).select(pl.col('a') < 'DO NOT COMPARE ME WITH NUMBERS')
```
Output
```
shape: (3, 1)
┌──────┐
│ a    │
│ ---  │
│ bool │
╞══════╡
│ true │
├╌╌╌╌╌╌┤
│ true │
├╌╌╌╌╌╌┤
│ true │
└──────┘
```

**Comment by SimonSchneider** [comment]
I guess the more useful side of this feature is if you have a string column which you filter by `<` number. There the type coercing is much more appealing, or are we considering that as "not allowed" either?
 
The title of the issue says "Comparison between strings and ints is allowed and should not be", I think there is ambiguity here that could be good to resolve. Is the problem that there is type coercing at all or is the problem specifically `number -> string` and not ie `string -> number`?

**Comment by cnpryer** [comment]
> It would be better to convert string to a number and then compare numbers.

But I don't think this is a good idea either. Why hand-hold here? Put the onus on the user to manage conversions for scenarios other than soft-compatible operations like *Number* with *Numbers* (ex: float and int).

**Comment by slonik-az** [comment]
@cnpryer I think the useful pattern is comparing dates to strings. You want to say something like `today < "2022-04-19"` assuming that the string will be coerced to the date. But I agree with you that with numbers this pattern does not improve ergonomics and should be considered an anti-pattern.

**Comment by cnpryer** [comment]
I may have a little more radical opinion here, but I'd even argue that shouldn't be coerced. In an explicit-first design, `today < "2022-04-19"` should throw `Cannot compare type Date with type String` or something. Why not put onus **entirely** on the user when you can and give them minimal, rational conversion based on standardized rationale like "Polars only implies like-types".

**Comment by ghuls** [comment]
> I guess the more useful side of this feature is if you have a string column which you filter by `<` number. There the type coercing is much more appealing, or are we considering that as "not allowed" either?
> 
> The title of the issue says "Comparison between strings and ints is allowed and should not be", I think there is ambiguity here that could be good to resolve. Is the problem that there is type coercing at all or is the problem specifically `number -> string` and not ie `string -> number`?

In my opinion, type coercing should not exist for number to string or string to number comparisons, but require an explicit cast.

**Comment by slonik-az** [comment]
Choosing between "no-implicit-coercion" and "sloppy-implicit-coercion" I will take "no-implicit-coercion" any time. Unfortunately, decades ago languages like Python and JavaScript popularized ill-conceived coercions  that many users learn to rely upon and expect from projects like polars.

**Comment by cnpryer** [comment]
iirc not even Pandas does this hand-holding. and Pandas implements a lot of indecisive stuff.

dates and strings
```py
import pandas as pd

s = pd.Series(["2022-04-14"])
today = pd.to_datetime("2022-04-19")
today > s
```
```
TypeError: '>' not supported between instances of 'str' and 'Timestamp'
```

ints and strings
```py
s = pd.Series(["7"])
8 > s
```
```
TypeError: '<' not supported between instances of 'str' and 'int'
```

**Comment by slonik-az** [comment]
Actually in pandas one can compare dates (specifically datetime periods) with strings.
```python
> ipython
Python 3.10.4 | packaged by conda-forge | (main, Mar 24 2022, 17:42:03) [Clang 12.0.1 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.2.0 -- An enhanced Interactive Python. Type '?' for help.
In [1]: import pandas as pd
In [2]: dr = pd.period_range(start='2022-01-01', end='2022-12-31', freq='M')
In [3]: dr
PeriodIndex(['2022-01', '2022-02', '2022-03', '2022-04', '2022-05', '2022-06',
             '2022-07', '2022-08', '2022-09', '2022-10', '2022-11', '2022-12'],
            dtype='period[M]')
In [4]: dr < "2022-07-15"
array([ True,  True,  True,  True,  True,  True, False, False, False,
       False, False, False])
```

**Comment by cnpryer** [comment]
But how is this not inconsistent design?

**Comment by slonik-az** [comment]
> But how is this not inconsistent design?

pandas `Timestamp` cannot be compared with `str` but datetime period can. I call it inconsistent design.

**Comment by cnpryer** [comment]
I'm not sure what this is about but just fyi you can view edits. I was asking because your comment originally read as a counter point.

I can remove my comment if you want me to since it's outdated. Didn't mean to cause confusion here. It looks like we're on the same page though.

**Comment by ritchie46** [comment]
Thanks for all the feedback. I think we most are on the same page that we should not coerce strings and numbers to their supertypes. I like putting type hygiëne on the user as well. Except for number number comparisons. 

I will make sure we error out on this.
